connect_ext_test.go

```diff
@@ -543,6 +543,43 @@ func TestTimeoutParsing(t *testing.T) {
 	assert.Nil(t, err)
 }
 
+func TestFailCodec(t *testing.T) {
+	t.Parallel()
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {})
+	server := httptest.NewServer(handler)
+	defer server.Close()
+	client := pingv1connect.NewPingServiceClient(
+		server.Client(),
+		server.URL,
+		connect.WithCodec(failCodec{}),
+	)
+	stream := client.CumSum(context.Background())
+	err := stream.Send(&pingv1.CumSumRequest{})
+	var connectErr *connect.Error
+	assert.NotNil(t, err)
+	assert.True(t, errors.As(err, &connectErr))
+	assert.Equal(t, connectErr.Code(), connect.CodeInternal)
+}
+
+func TestContextError(t *testing.T) {
+	t.Parallel()
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {})
+	server := httptest.NewServer(handler)
+	defer server.Close()
+	client := pingv1connect.NewPingServiceClient(
+		server.Client(),
+		server.URL,
+	)
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	stream := client.CumSum(ctx)
+	err := stream.Send(nil)
+	var connectErr *connect.Error
+	assert.NotNil(t, err)
+	assert.True(t, errors.As(err, &connectErr))
+	assert.Equal(t, connectErr.Code(), connect.CodeCanceled)
```

Comment:
Can we add an `assert.False(t, connect.IsWireError(err))` too? The client should be short-circuiting and returning an error before even writing anything to the network.